### PR TITLE
fix(metrics): prevent reset-password from clobbering mixed-in events

### PR DESCRIPTION
**Pull request by philbooth**
Fixes mozilla/fxa-amplitude-send#65.

The reset password view was clobbering the events that it was mixing in from the flow events mixin, by assigning directly to `this.events`.  This change sorts that out, and adds some tests that all of the expected events are registered for.

@mozilla/fxa-devs r?

### CODE REVIEW

File: app/scripts/views/reset_password.js [hunk 1]
```diff
@@ -11,14 +11,17 @@ define(function (require, exports, module) {
...
   const PasswordResetMixin = require('./mixins/password-reset-mixin');
-  const preventDefaultThen = require('./base').preventDefaultThen;
   const ServiceMixin = require('./mixins/service-mixin');
   const Session = require('../lib/session');
   const Template = require('templates/reset_password.mustache');
 
   const t = (msg) => msg;
 
-  class ResetPasswordView extends FormView {
+  const ResetPasswordView = FormView.extend({
```
Comment (philbooth) [inline app/scripts/views/reset_password.js]: I could probably also have done this without ditching `class` I guess, by setting `events` on `ResetPasswordView.prototype` instead. I opted to do it this way to be consistent with the other views, but I'm happy to change it if we'd prefer to gradually move in the direction of `class` syntax.

File: app/scripts/views/reset_password.js [hunk 2]
```diff
@@ -11,14 +11,17 @@ define(function (require, exports, module) {
...
+    events: {
+      'click .remember-password': '_rememberPassword'
```
Comment (shane-tomlinson) [inline app/scripts/views/reset_password.js]: Do we still need the `preventDefaultThen` here?

File: app/scripts/views/reset_password.js [hunk 3]
```diff
@@ -27,22 +30,13 @@ define(function (require, exports, module) {
       // address clears the email field in the formPrefill model if
       // the user doesn't enter an address. See comment in beforeDestroy
       this._formPrefill = options.formPrefill;
-
-      this.events = {
-        'click .remember-password': preventDefaultThen('_rememberPassword')
-      };
-
-      super.initialize(options);
```
Comment (shane-tomlinson) [inline app/scripts/views/reset_password.js]: To future proof ourselves, I think there should be a `FormView.prototype.initialize.call(this, options);` here.

File: app/scripts/views/reset_password.js [hunk 4]
```diff
@@ -27,22 +30,13 @@ define(function (require, exports, module) {
...
-    }
-
+    },
 
     setInitialContext (context) {
-      super.setInitialContext(context);
```
Comment (shane-tomlinson) [inline app/scripts/views/reset_password.js]: To future proof ourselves, I think there should be a `FormView.prototype.setInitialContext.call(this, context);` here.

File: app/scripts/views/reset_password.js [hunk 5]
```diff
@@ -54,9 +48,7 @@ define(function (require, exports, module) {
             this.model.set('error', err);
           });
       }
-
-      return super.beforeRender();
```
Comment (shane-tomlinson) [inline app/scripts/views/reset_password.js]: To future proof ourselves, I think there should be a `return FormView.prototype.beforeRender.call(this);` here.

File: app/scripts/views/reset_password.js (hunk 1 above)
Comment (shane-tomlinson) [reply app/scripts/views/reset_password.js]: I've been trying to figure out how to handle adding events and class syntax. Over in #6273 I have added the [transform-class-properties](https://github.com/mozilla/fxa-content-server/pull/6273/files#diff-11e9f7f953edc64ba14b0cc350ae7b9dR113) babel plugin which allows us to use the new [class properties syntax](https://babeljs.io/docs/en/babel-plugin-transform-class-properties/).

Using class properties we _could_ do something like:

```js
class ResetPasswordView extends FormView {
  events = {
    event_name: 'function_name'
  };

  constructor (options) {
    super(options);
    ...
  }
  ...
}
```

That's decent looking but I suspect it's broken with mixins semantically that compiles down to something similar to what you are fixing here. The way the class properties work, they are not set on the prototype as one would expect, rather the constructor is called and THEN the class properties are set, just like the version being fixed. They are not "class properties" but rather "instance properties". This implies everything will be broken using class properties if `events` are needed.

I'm sure there is some hackery to get around that, I just haven't figured it out yet.

File: app/scripts/views/reset_password.js (hunk 2 above)
Comment (philbooth) [reply app/scripts/views/reset_password.js]: Not as far as I can tell, unless I'm missing something?

Comment (philbooth) [reply app/scripts/views/reset_password.js]: I tell a lie, yes we do.

**Pull request merged**